File: services/gemini_service.py
import google.generativeai as genai
from typing import Optional, List, Dict, Any
import base64
import os

API_KEY = os.getenv("GEMINI_API_KEY", "")

BASE_CLEAN_TEXT_INSTRUCTION = """
IMPORTANT: Do not use Markdown symbols like #, *, **, or _ in your response. 
Do not use labels like "The Text:", "The Explanation:", or "Explanation:". 
Present information in a clean, natural reading style. 
Use double line breaks between sections for clarity. 
If you need to list items, use simple numbering like 1. 2. 3. without any special characters around the numbers.
"""

def get_socratic_instruction(is_socratic: bool) -> str:
    base = "You are a Socratic Tutor. Never give the direct answer. Provide guidance and ask questions to lead the student to the answer." if is_socratic else "You are a helpful, direct study buddy."
    return base + BASE_CLEAN_TEXT_INSTRUCTION

def explain_concept(concept: str, image_base64: Optional[str] = None, socratic: bool = False) -> Dict[str, Any]:
    """Explain a concept using Gemini API"""
    try:
        model = genai.GenerativeModel("gemini-2.0-flash")
        
        prompt = f"Explain this: {concept}"
        
        if image_base64:
            # Handle image if provided
            image_data = base64.standard_b64decode(image_base64.split(',')[1] if ',' in image_base64 else image_base64)
            response = model.generate_content([prompt, {"mime_type": "image/jpeg", "data": image_data}])
        else:
            response = model.generate_content(prompt)
        
        return {
            "text": response.text,
            "sources": []
        }
    except Exception as e:
        return {
            "text": f"Error: {str(e)}",
            "sources": []
        }

def summarize_text(text: str, length: str = "Medium") -> str:
    """Summarize text using Gemini API"""
    try:
        model = genai.GenerativeModel("gemini-2.0-flash")
        
        length_instruction = {
            "Brief": "Provide a very concise summary (2-3 sentences)",
            "Medium": "Provide a moderate summary (1-2 paragraphs)",
            "Detailed": "Provide a detailed summary (3-4 paragraphs)"
        }
        
        prompt = f"{length_instruction.get(length, 'Provide a summary')} of the following text:\n\n{text}"
        response = model.generate_content(prompt)
        return response.text
    except Exception as e:
        return f"Error: {str(e)}"

def generate_quiz(topic: str, num_questions: int = 5, difficulty: str = "Medium") -> List[Dict[str, Any]]:
    """Generate quiz questions using Gemini API"""
    try:
        model = genai.GenerativeModel("gemini-2.0-flash")
        
        prompt = f"""Generate {num_questions} multiple-choice quiz questions about {topic} at {difficulty} difficulty level.
        
        For each question, provide:
        1. Question text
        2. Four options (A, B, C, D)
        3. Correct answer
        4. Explanation
        
        Format as JSON array."""
        
        response = model.generate_content(prompt)
        import json
        try:
            return json.loads(response.text)
        except:
            return [{"question": response.text, "options": [], "correct": "A", "explanation": ""}]
    except Exception as e:
        return [{"question": f"Error: {str(e)}", "options": [], "correct": "A", "explanation": ""}]

def generate_flashcards(topic: str, num_cards: int = 10) -> List[Dict[str, str]]:
    """Generate flashcards using Gemini API"""
    try:
        model = genai.GenerativeModel("gemini-2.0-flash")
        
        prompt = f"""Generate {num_cards} flashcards for studying {topic}.
        
        For each flashcard, provide:
        - front: The question or prompt
        - back: The answer or explanation
        
        Format as JSON array."""
        
        response = model.generate_content(prompt)
        import json
        try:
            return json.loads(response.text)
        except:
            return [{"front": topic, "back": response.text}]
    except Exception as e:
        return [{"front": topic, "back": f"Error: {str(e)}"}]

def analyze_document(file_content: str, analysis_type: str = "Summary") -> str:
    """Analyze document content using Gemini API"""
    try:
        model = genai.GenerativeModel("gemini-2.0-flash")
        
        prompts = {
            "Summary": f"Provide a comprehensive summary of:\n\n{file_content}",
            "Key Points": f"List the main key points from:\n\n{file_content}",
            "Quiz Generation": f"Generate 5 quiz questions based on:\n\n{file_content}",
            "Explanation": f"Provide a detailed explanation of the concepts in:\n\n{file_content}"
        }
        
        prompt = prompts.get(analysis_type, f"Analyze:\n\n{file_content}")
        response = model.generate_content(prompt)
        return response.text
    except Exception as e:
        return f"Error: {str(e)}"

def generate_mnemonics(concept: str, mnemonic_type: str = "Acronym") -> str:
    """Generate mnemonics for a concept using Gemini API"""
    try:
        model = genai.GenerativeModel("gemini-2.0-flash")
        
        type_prompts = {
            "Acronym": f"Create an acronym mnemonic for remembering {concept}",
            "Method of Loci": f"Create a Method of Loci (memory palace) for {concept}",
            "Rhyme": f"Create a rhyming mnemonic for {concept}",
            "Story": f"Create a memorable story to remember {concept}",
            "Association": f"Create word associations to remember {concept}"
        }
        
        prompt = type_prompts.get(mnemonic_type, f"Create a mnemonic for {concept}")
        response = model.generate_content(prompt)
        return response.text
    except Exception as e:
        return f"Error: {str(e)}"

def generate_story(topic: str, style: str = "Educational", audience: str = "Adults") -> str:
    """Generate an educational story using Gemini API"""
    try:
        model = genai.GenerativeModel("gemini-2.0-flash")
        
        prompt = f"Write a {style} story about {topic} for {audience}. Make it engaging and educational."
        response = model.generate_content(prompt)
        return response.text
    except Exception as e:
        return f"Error: {str(e)}"

File: services/geminiService.ts

import { GoogleGenAI, Type, GenerateContentResponse } from "@google/genai";
import { QuizQuestion, Flashcard, SummaryResult, ChatMessage, MnemonicItem, StoryResult } from "../types";

const API_KEY = "AIzaSyDRPnG-GZm_H234VaGFheeY39FqnpjeQ4Y"|| "";

const BASE_CLEAN_TEXT_INSTRUCTION = `
IMPORTANT: Do not use Markdown symbols like #, *, **, or _ in your response. 
Do not use labels like "The Text:", "The Explanation:", or "Explanation:". 
Present information in a clean, natural reading style. 
Use double line breaks between sections for clarity. 
If you need to list items, use simple numbering like 1. 2. 3. without any special characters around the numbers.`;

const getSocraticInstruction = (isSocratic: boolean) => {
  const base = isSocratic 
    ? `You are a Socratic Tutor. Never give the direct answer. Provide guidance and ask questions to lead the student to the answer.` 
    : `You are a helpful, direct study buddy.`;
  
  return base + BASE_CLEAN_TEXT_INSTRUCTION;
};

export const explainConcept = async (concept: string, imageBase64?: string, isSocratic: boolean = false) => {
  const ai = new GoogleGenAI({ apiKey: API_KEY });
  const parts: any[] = [{ text: `Explain this: ${concept}` }];
  
  if (imageBase64) {
    parts.push({
      inlineData: {
        mimeType: 'image/jpeg',
        data: imageBase64.split(',')[1] || imageBase64,
      },
    });
  }

  const response: GenerateContentResponse = await ai.models.generateContent({
    model: 'gemini-3-pro-preview',
    contents: { parts },
    config: {
      systemInstruction: getSocraticInstruction(isSocratic),
      tools: [{ googleSearch: {} }],
    },
  });
  
  const sources = response.candidates?.[0]?.groundingMetadata?.groundingChunks?.map((chunk: any) => ({
    title: chunk.web?.title || "Reference",
    uri: chunk.web?.uri || "#"
  })) || [];

  return {
    text: response.text,
    sources
  };
};

export const chatWithDocument = async (history: ChatMessage[], documentContent: string, isSocratic: boolean = false) => {
  const ai = new GoogleGenAI({ apiKey: API_KEY });
  const lastUserMsg = history[history.length - 1].text;
  
  const response = await ai.models.generateContent({
    model: 'gemini-3-flash-preview',
    contents: [
      { role: 'user', parts: [{ text: `CONTEXT FROM DOCUMENT:\n${documentContent}\n\nUSER QUESTION: ${lastUserMsg}` }] }
    ],
    config: {
      systemInstruction: getSocraticInstruction(isSocratic) + " Base your responses strictly on the provided document context if relevant. " + BASE_CLEAN_TEXT_INSTRUCTION,
    }
  });

  return response.text;
};

export const summarizeNotes = async (text: string): Promise<SummaryResult> => {
  const ai = new GoogleGenAI({ apiKey: API_KEY });
  const response = await ai.models.generateContent({
    model: 'gemini-3-flash-preview',
    contents: `Summarize the following study notes. Provide a concise title, a brief summary paragraph, and a list of key takeaways. ${BASE_CLEAN_TEXT_INSTRUCTION} \n\n${text}`,
    config: {
      responseMimeType: "application/json",
      responseSchema: {
        type: Type.OBJECT,
        properties: {
          title: { type: Type.STRING },
          summary: { type: Type.STRING },
          keyPoints: {
            type: Type.ARRAY,
            items: { type: Type.STRING }
          }
        },
        required: ["title", "summary", "keyPoints"]
      }
    }
  });

  return JSON.parse(response.text || "{}");
};

export const generateQuiz = async (topicOrContext: string): Promise<QuizQuestion[]> => {
  const ai = new GoogleGenAI({ apiKey: API_KEY });
  const response = await ai.models.generateContent({
    model: 'gemini-3-flash-preview',
    contents: `Generate 5 challenging multiple choice questions based on the following content: ${topicOrContext}. Each question must have exactly 4 options and one correct answer index (0-3). Include a helpful explanation. ${BASE_CLEAN_TEXT_INSTRUCTION}`,
    config: {
      responseMimeType: "application/json",
      responseSchema: {
        type: Type.ARRAY,
        items: {
          type: Type.OBJECT,
          properties: {
            question: { type: Type.STRING },
            options: {
              type: Type.ARRAY,
              items: { type: Type.STRING }
            },
            correctAnswer: { type: Type.INTEGER },
            explanation: { type: Type.STRING }
          },
          required: ["question", "options", "correctAnswer", "explanation"]
        }
      }
    }
  });

  return JSON.parse(response.text || "[]");
};

export const generateFlashcards = async (topicOrNotes: string): Promise<Flashcard[]> => {
  const ai = new GoogleGenAI({ apiKey: API_KEY });
  const response = await ai.models.generateContent({
    model: 'gemini-3-flash-preview',
    contents: `Generate 8 study flashcards (front-back style) for the following topic or text: ${topicOrNotes}. ${BASE_CLEAN_TEXT_INSTRUCTION}`,
    config: {
      responseMimeType: "application/json",
      responseSchema: {
        type: Type.ARRAY,
        items: {
          type: Type.OBJECT,
          properties: {
            front: { type: Type.STRING },
            back: { type: Type.STRING }
          },
          required: ["front", "back"]
        }
      }
    }
  });

  return JSON.parse(response.text || "[]");
};

export const generateMnemonics = async (text: string): Promise<MnemonicItem[]> => {
  const ai = new GoogleGenAI({ apiKey: API_KEY });
  const response = await ai.models.generateContent({
    model: 'gemini-3-flash-preview',
    contents: `Generate 4 creative mnemonics to help remember key terms in the following text. ${BASE_CLEAN_TEXT_INSTRUCTION} \n\n${text}`,
    config: {
      responseMimeType: "application/json",
      responseSchema: {
        type: Type.ARRAY,
        items: {
          type: Type.OBJECT,
          properties: {
            type: { type: Type.STRING },
            phrase: { type: Type.STRING },
            explanation: { type: Type.STRING }
          },
          required: ["type", "phrase", "explanation"]
        }
      }
    }
  });
  return JSON.parse(response.text || "[]");
};

export const generateStory = async (text: string, theme: string): Promise<StoryResult> => {
  const ai = new GoogleGenAI({ apiKey: API_KEY });
  const response = await ai.models.generateContent({
    model: 'gemini-3-flash-preview',
    contents: `Write an engaging story explaining the concepts in the following text. Theme: "${theme}". ${BASE_CLEAN_TEXT_INSTRUCTION} \n\nSTUDY TEXT: ${text}`,
    config: {
      responseMimeType: "application/json",
      responseSchema: {
        type: Type.OBJECT,
        properties: {
          title: { type: Type.STRING },
          content: { type: Type.STRING }
        },
        required: ["title", "content"]
      }
    }
  });
  return JSON.parse(response.text || "{}");
};
